neoload/commands/fastfail.py

```diff
@@ -0,0 +1,155 @@
+@click.option("--force", 'force', is_flag=True, default=True, help="Doesn't wait the end of test")
+@click.option("--max-failure", 'max_failure', type=int, default=0, help="Max SLA failure threshold; default is zero")
+def cli(command, name, stop, force, max_failure): #, max_occurs):
+    """Fails if certain conditions are met, such as per-run SLAs failed % of time"""
+    if not command:
+        tools.system_exit({'message': "command is mandatory. Please see neoload fastfail --help", 'code': 2})
+        return
+
+    if max_failure < 0 or max_failure > 100:
+        tools.system_exit({'message': "--max-failure percentage tolerance must be between 0 and 100", 'code': 2})
+        return
+
+
+    if command == "slas":
+        monitor_loop(name, stop, force, max_failure)
+
+    else:
+        tools.system_exit({'message': "Invalid command. Please see neoload fastfail --help", 'code': 2})
+
+def monitor_loop(name, stop, force, max_failure):
+    dt_started = datetime.now()
+    print('fastfail started: ' + str(dt_started))
+    dt_current = dt_started
+
+    if sys.stdin.isatty():
+        sys.stdout = Unbuffered(sys.stdout)
+
+    is_initializing = False
+    is_running = False
+    has_exited = False
+    fails = []
+    msg = ""
+    exit_code = 0
+    while (abs(dt_current-dt_started).seconds / 60) < 10:
+        datas = test_results.get_sla_data_by_name_or_id(name)
+
+        partial_intervals = list(filter(lambda x: x['status']=='FAILED',datas['sla_interval']))
+        failed_intervals = list(filter(lambda x: x['failed']>=max_failure,partial_intervals))
+
+        fails = []
+        fails.extend(list(filter(lambda x: x['status']=='FAILED',datas['sla_global'])))
+        fails.extend(list(filter(lambda x: x['status']=='FAILED',datas['sla_test'])))
+        fails.extend(failed_intervals)
+
+        status = datas['result']['status']
+        quality_status = datas['result']['qualityStatus']
+
+        msg = "{0} hard SLA failures, test status {1},{2}".format(len(fails), status, quality_status)
+        if len(partial_intervals) > 0 and len(failed_intervals) < 1:
+            msg += "\nSome SLAs failed, but were not above max-failure of {0}% so did not count as error.".format(max_failure)
+
+        outcomes = process_state(datas,fails,stop,force,is_initializing,is_running,msg)
+
+        final_run = has_exited and outcomes["has_exited"]
+        has_exited = outcomes["has_exited"]
+        is_initializing = outcomes["is_initializing"]
+        is_running = outcomes["is_running"]
+        exit_code = 0 if datas['result']['qualityStatus']=="PASSED" else 1
+
+        if final_run:
+            break
+        elif (len(fails) > 0 or len(partial_intervals) > 0):
+            displayer.__print_sla(datas['sla_global'], datas['sla_test'], datas['sla_interval'])
+        else:
+            printif(sys.stdin.isatty() and not has_exited, '.', end = '')
+            time.sleep(5)
```

Comment:
The sleep time of 5s is very short.
Are you sure that if people launch for example 3 tests at the same time on Neoload Web, the API limit of 300calls per minute won't be reached ?
The function "get_sla_data_by_name_or_id" makes 5 calls to the API...

Reply:
Resolved with 37a6a7e